pkg/apis/core/v1alpha1/extensions_federatedobject.go

```diff
@@ -0,0 +1,120 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+This file may have been modified by The KubeAdmiral Authors
+("KubeAdmiral Modifications"). All KubeAdmiral Modifications
+are Copyright 2023 The KubeAdmiral Authors.
+*/
+
+package v1alpha1
+
+import (
+	"reflect"
+	"sort"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+// Placement extensions
+
+// GetPlacementUnion returns the union of all clusters listed under the Placement field of the FederatedObject.
+func (o *FederatedObject) GetPlacementUnion() sets.Set[string] {
+	set := sets.New[string]()
+	for _, placement := range o.Spec.Placements {
+		for _, cluster := range placement.Clusters {
+			set.Insert(cluster.Name)
+		}
+	}
+	return set
+}
+
+// GetControllerPlacement returns the slice containing all the ClusterPlacements from a given controller. Returns nil if
+// the controller is not present.
+func (o *FederatedObject) GetControllerPlacement(controller string) []ClusterPlacement {
+	for _, placement := range o.Spec.Placements {
+		if placement.Controller == controller {
+			return placement.Clusters
+		}
+	}
+	return nil
+}
+
+// SetControllerPlacement sets the ClusterPlacements for a given controller. If clusterNames is nil or empty, the previous
+// placement for the given controller will be deleted. Returns a bool indicating if the FederatedObject has changed.
+func (o *FederatedObject) SetControllerPlacement(controller string, clusterNames []string) bool {
+	if len(clusterNames) == 0 {
+		return o.DeleteControllerPlacement(controller)
+	}
+
+	newPlacement := PlacementWithController{
+		Controller: controller,
+		Clusters:   make([]ClusterPlacement, len(clusterNames)),
```

Comment:
Can extract this to a separate variable and construct `PlacementWithController` later? Otherwise it is easy to accidentally insert a line after line 65 that uses `newPlacement` assuming it is a valid value as a whole.